client/src/explorer_common/state_and_memoizing.js

```diff
@@ -71,6 +71,40 @@ function root_reducer(state = initial_root_state, action) {
       }
     }
 
+    case "expand_all": {
+      const { root } = payload;
+      const { userExpanded: oldExpanded } = state;
+
+      const get_ids_to_update = ({ isExpanded, id, children }) => [
+        !isExpanded ? id : [],
+        _.map(children, get_ids_to_update),
+      ];
+      const ids_to_update = _.flattenDeep(get_ids_to_update(root));
+
+      return {
+        ...state,
+        userExpanded: oldExpanded.concat(ids_to_update),
+        userCollapsed: [],
+      };
+    }
+
+    case "collapse_all": {
+      const { root } = payload;
+      const { userCollapsed: oldCollapsed } = state;
+
+      const get_ids_to_update = ({ isExpanded, id, children }) => [
+        isExpanded ? id : [],
+        _.map(children, get_ids_to_update),
+      ];
+      const ids_to_update = _.flattenDeep(get_ids_to_update(root));
+
+      return {
+        ...state,
+        userExpanded: [],
+        userCollapsed: oldCollapsed.concat(ids_to_update),
+      };
+    }
```

Comment:
I'm thinking this block can be cleaned up with a helper function to avoid repeating a lot of shared code between expanding and collapsing all.

Reply:
Sounds good to me!